### ISSUE TITLE
Unexpected reportMatchNotExhaustive for exhaustive match on TypedDict union

### ISSUE THREAD

**Issue by alessiogaldy** [microsoft/pyright#4868]
**Describe the bug**
Exhaustive match statement on union of `TypedDict` is flagged by reportMatchNotExhaustive

**To Reproduce**
```python
class IntValue(TypedDict):
    type: Literal["Int"]
    int_value: int


class StrValue(TypedDict):
    type: Literal["Str"]
    str_value: str

Value = IntValue | StrValue

def print_value(value: Value) -> None:
    match value:
        case {"type": "Int"}:
            print(value["int_value"])
        case {"type": "Str"}:
            print(value["str_value"])
```

**Expected behavior**
match statement shouldn't be flagged as non-exhaustive because all variants are checked 

**VS Code extension or command-line**
* VS Code pylance extension version: 2023.3.40
* pyright cli version: 1.1.301

**Comment by erictraut** [comment]
Pyright doesn't currently perform any type narrowing in the negative (fall-through) case for mapping patterns. Mypy doesn't do this either. So, this is by design currently.

I think it's reasonable (and sound, from a type perspective) to support this for unions of TypeDicts that are discriminated by a literal field. Pyright does support this form of narrowing in a comparable `if` statement:

```python
def print_value2(value: Value) -> None:
    if value["type"] == "Int":
        print(value["int_value"])
    elif value["type"] == "Str":
        print(value["str_value"])
    else:
        assert_never(value)
```

I'll investigate adding this support for mapping patterns in a `case` statement.

**Comment by alessiogaldy** [comment]
Thank you for taking a look! I didn't realized this was by design.

**Comment by erictraut** [comment]
This support will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.302, which I just published. It will also be included in a future release of pylance.
